pool/pool.go

```diff
@@ -30,284 +31,576 @@ import (
+	ObjectSearchInit(context.Context, sdkClient.PrmObjectSearch) (*sdkClient.ObjectListReader, error)
+	SessionCreate(context.Context, sdkClient.PrmSessionCreate) (*sdkClient.ResSessionCreate, error)
+}
+
+// InitParameters contains values used to initialize connection Pool.
+type InitParameters struct {
+	key                       *ecdsa.PrivateKey
+	logger                    *zap.Logger
+	nodeConnectionTimeout     time.Duration
+	nodeRequestTimeout        time.Duration
+	clientRebalanceInterval   time.Duration
+	sessionTokenThreshold     time.Duration
+	sessionExpirationDuration uint64
+	nodeParams                []NodeParam
+
+	clientBuilder func(endpoint string) (client, error)
+}
+
+// SetKey specifies default key to be used for the protocol communication by default.
+func (x *InitParameters) SetKey(key *ecdsa.PrivateKey) {
+	x.key = key
+}
+
+// SetLogger specifies logger.
+func (x *InitParameters) SetLogger(logger *zap.Logger) {
+	x.logger = logger
+}
+
+// SetNodeConnectionTimeout specifies the timeout for connection to be established.
+func (x *InitParameters) SetNodeConnectionTimeout(timeout time.Duration) {
+	x.nodeConnectionTimeout = timeout
+}
+
+// SetNodeRequestTimeout specifies the timeout for request to node to decide if it is alive.
+func (x *InitParameters) SetNodeRequestTimeout(timeout time.Duration) {
+	x.nodeRequestTimeout = timeout
+}
+
+// SetClientRebalanceInterval specifies the interval for updating nodes health status.
+func (x *InitParameters) SetClientRebalanceInterval(interval time.Duration) {
+	x.clientRebalanceInterval = interval
+}
+
+// SetSessionTokenThreshold specifies the max session token life time for PutObject operation.
+func (x *InitParameters) SetSessionTokenThreshold(threshold time.Duration) {
+	x.sessionTokenThreshold = threshold
+}
+
+// SetSessionExpirationDuration specifies the session token lifetime in epochs.
+func (x *InitParameters) SetSessionExpirationDuration(expirationDuration uint64) {
+	x.sessionExpirationDuration = expirationDuration
+}
+
+// SetNodeParams specifies information about the nodes to which you want to connect.
+func (x *InitParameters) SetNodeParams(nodeParams []NodeParam) {
+	x.nodeParams = nodeParams
+}
+
+type rebalanceParameters struct {
+	nodesParams               []*nodesParam
+	nodeRequestTimeout        time.Duration
+	clientRebalanceInterval   time.Duration
+	sessionExpirationDuration uint64
+}
+
+type nodesParam struct {
 	priority  int
 	addresses []string
 	weights   []float64
 }
 
+// NodeParam groups parameters of remote node.
 type NodeParam struct {
 	priority int
 	address  string
 	weight   float64
 }
 
-// Builder is an interim structure used to collect node addresses/weights and
-// build connection Pool subsequently.
-type Builder struct {
-	nodeParams []NodeParam
+// NewNodeParam creates NodeParam using parameters.
+func NewNodeParam(priority int, address string, weight float64) NodeParam {
```

Comment:
For me, this function looks convenient only as long as there are 3 node parameters. Mentally increasing them to, say, 6, it loses all meaning.

Reply:
For me it's too convenient to use:
```
prm := NewNodeParam(1, "peer", 1)
```
instead of:
```
var prm NodeParam
prm.SetPriority(1)
prm.SetAddress("peer")
prm.SetWeight(1)
```

I believe we need something like that to prevent copypasting the same funtions in application code.

Reply:
I agree with the convenience of the function, but how we'll deal with potentially new parameters? Smth like that:
```
prm := NewNodeParam(1, "peer", 2)
prm.SetSmthElse(val)
```
I mean that we won't be able to add the parameter to the function JFYI.

Btw it should be implemented just like you wrote:
```
func NewNodeParam(priority int, address string, weight float64) (prm NodeParam) {
  prm.SetPriority(priority)
  prm.SetAddress(address)
  prm.SetWeight(weight)

  return
}
```
in order to not re-implement setters.